efi/shim_load_handler.go

```diff
@@ -0,0 +1,341 @@
+	if err != nil {
+		return nil, xerrors.Errorf("cannot read vendor DB: %w", err)
+	}
+	vendorDbName := efi.VariableDescriptor{Name: shimName, GUID: shimGuid}
+
+	switch {
+	case format == shimVendorCertIsDb:
+		flags |= shimVendorCertContainsDb
+
+		// This is surely wrong
+		vendorDbName = efi.VariableDescriptor{
+			Name: shimVendorDbName,
+			GUID: efi.ImageSecurityDatabaseGuid}
+	case format == shimVendorCertIsX509 && ver.Compare(mustParseShimVersion("15.7")) >= 0:
+		// Because of https://github.com/rhboot/shim/commit/092c2b2bbed950727e41cf450b61c794881c33e7,
+		// shim marks verification events that use the built-in vendor cert with the MokListRT name
+		// and includes the shim GUID as the signature's owner GUID.
+		// 15.7 introduced this bug
+		vendorDbName = efi.VariableDescriptor{
+			Name: shimMokListRTName,
+			GUID: shimGuid}
+		vendorDb[0].Signatures[0].Owner = shimGuid
+
+		// set this because the bug also changes the measurement format - it measures the entire
+		// EFI_SIGNATURE_DATA structure rather than just the certificate.
+		flags |= shimVendorCertContainsDb
+	}
+
+	// Newer versions of shim carry a payload which it writes to a BS+NV variable to set the SBAT
+	// revocation policy (SbatLevel). Note that shim uses "SBAT policy" for another meaning,
+	// which makes it a bit confusing here.
+	//
+	// If the section doesn't exist, then the payload must be supplied externally.
+	var sbatLevel shimSbatLevel
+	if c.sbatLevel != nil {
+		sbatLevel = *c.sbatLevel
+	} else if shim.HasSbatLevelSection() {
+		sbatLevel, err = shim.ReadSbatLevel()
+		if err != nil {
+			return nil, xerrors.Errorf("cannot read SbatLevel from shim: %w", err)
+		}
+	}
+
+	if authoritySet, ok := rules.(secureBootAuthoritySet); ok {
+		// Add built-in vendor certs from the rule set that constructed us so that
+		// rules that are based around a secure boot namespace recognize images
+		// signed by these vendor certs as part of the same namespace.
+		var certs []*x509.Certificate
+		for _, esl := range vendorDb {
+			if esl.Type != efi.CertX509Guid {
+				continue
+			}
+			if len(esl.Signatures) == 0 {
+				continue
+			}
+
+			cert, err := x509.ParseCertificate(esl.Signatures[0].Data)
+			if err != nil {
+				return nil, xerrors.Errorf("cannot parse vendor cert: %w", err)
+			}
+			certs = append(certs, cert)
+		}
+		authoritySet.AddAuthorities(certs...)
+	}
+
+	return &shimLoadHandler{
+		Flags: flags,
+		VendorDb: &secureBootDB{
+			Name:     vendorDbName,
+			Contents: vendorDb,
+		},
+		SbatLevel: sbatLevel}, nil
+}
+
+func newShimLoadHandler(rules imageRuleSet, image peImageHandle) (imageLoadHandler, error) {
+	return newShimLoadHandlerConstructor().New(rules, image)
+}
+
+// MeasureImageStart implements imageLoadHandler.MeasureImageStart.
+func (h *shimLoadHandler) MeasureImageStart(ctx pcrBranchContext) error {
+	// Update the context for this branch
+	ctx.ShimContext().Flags = h.Flags
+	ctx.ShimContext().VendorDb = h.VendorDb
```

Comment:
this is a bit strange, as I remarked the same for `ctx.FwContext().Db = &secureBootDB{Name: Db, Contents: db}` in fw_load_handler.go. I think the pcrBranchContext should start without ShimContext and FwContext and have way to set a full one of those, or the context need some sort of Init or similar function that take the values they need to be initialized all together instead of set one by one.  It's also unclear whether VendorDB should be public on shimContext or a method that panics more cleanly if the context/vendorDb has not been initialised yet..